Enahnced data type support

RichardD2 opened the pull request:
This update adds support for model properties of type `SecureString` and `byte[]` in addition to `string`.

It also supports storing the encrypted values in `varbinary` columns instead of Base64-encoded `varchar` columns.

I've also added basic migration support for moving between encryption providers (including to and from unencrypted data).

And finally, I've restored the "fixed IV" option on the `AesProvider` for cases where the encrypted column needs to be searched with LINQ, or where the IV being changed on every update causes problems.

Eastrall commented:
Thanks for contributing to the project @RichardD2 ! I got to say, the PR's decription looks really nice! I am going find some time to review this during this weekend.

I noticed that you included the `MigrationEncryptionProvider` as you suggested in PR #15, it simplifies my original implementation on this PR! Nice. 😄

Eastrall commented:
I went throught the changes and first of all, I have some feedback about the coding style, mostly about the curly braces `{ }` in general.
Most of the one-line condition doesn't have braces and are written in only one line, which, imo, makes it difficult to read:

```csharp
if (converter is null) throw new ArgumentNullException(nameof(converter));
if (encoder is null) throw new ArgumentNullException(nameof(encoder));
```
Would be:
```csharp
if (converter is null)
{
    throw new ArgumentNullException(nameof(converter));
}
if (encoder is null)
{
    throw new ArgumentNullException(nameof(encoder));
}
```
Also, there are several internal classes in the same file: `ConverterBuilder`, it would be better for readability to split these classes into multiple files. Also, the `ConverterBuilder` could be `internal` since it's using only for internal work.

If I understand correctly the new `IEncryptionProvider` behavior, its now using generics to store and load any kind of data (`string`, `byte[]`, etc...) ; when storing data, the actual `TModel` value is passed into a **converter** that will transform the given data into a `byte[]`. Which will be passed to the **encoder** with the previously **converted** `byte[]` data, wrapped into a `Stream`, to return the **real encoded** data that will be stored into the database. Did I get this correctly?

I believe we could make some refactorings to make the code smaller and lighter. Maybe we could start by giving some code space to improve readability .

Nonetheless, the code looks good and seem to work. 

We should add some extra properties to the `AesProviderTest` to test the `byte[]` property encoding/decoding.

RichardD2 commented:
Your summary of the new `IEncryptionProvider` behaviour is correct.

I've added an `.editorconfig` file, fixed the single-line braces style, and restored your preferred space indent, which should hopefully clean up the diff.

I've also refactored the `ConverterBuilder` class to use `readonly struct`s, which should reduce the overhead of building a converter. The naming might seem a little odd - three classes called `ConverterBuilder`, which vary by the number of type parameters - but I couldn't think of a cleaner name for them.

I've left the classes `public` so that they can support additional model and store types. The `ConverterBuilder.From` and `ConverterBuilder.To` extension methods allow you to pass in delegates to encode and decode the custom type. For example:

```c#
ValueConverter<long, byte[]> int64Converter = encryptionProvider
    .From((long x) => BitConverter.GetBytes(x), (Stream s) => BitConverter.ToInt64(ReadAllBytes(s)))
    .ToBinary();
```

Eastrall commented on src/EntityFrameworkCore.DataEncryption/Internal/ConverterBuilder`1.cs:
```diff
@@ -0,0 +1,37 @@
+            Encoder = encoder;
+        }
+
+        private IEncryptionProvider EncryptionProvider { get; }
```
If the properties are not exposed, we could just put them as `private readonly` fields in the struct.

RichardD2 replied:
Good idea. 👍

Eastrall commented on src/EntityFrameworkCore.DataEncryption/Internal/ConverterBuilder`2.cs:
```diff
@@ -0,0 +1,69 @@
+            Encoder = encoder;
+        }
+
+        private ConverterBuilder<TModelType> ModelType { get; }
```
Same as above.

Eastrall commented:
Well, I guess the PR seems correct. Can you rebase your PR, it seems that there is a conflict with `EntityFrameworkCore.DataEncryption.sln`

RichardD2 commented:
I think that should be done now - rebasing in Visual Studio is a bit of a pain.

Eastrall reviewed (approved):
Looks good to me. Let's merge this PR! 😄

Eastrall commented:
Nuget package will be updated shortly.
